internal/cli/cluster/engine.go

```diff
@@ -0,0 +1,248 @@
+		return nil, err
+	}
+
+	// build extra schema in sub chart, now, we only support one sub chart
+	for _, subChart := range c.Dependencies() {
+		eSchema.SubChartName = subChart.Name()
+		eSchema.SubSchema, err = buildSchema(subChart.Schema)
+		if err != nil {
+			return nil, err
+		}
+		break
+	}
+
+	// if sub chart has alias, we should use alias instead of chart name
+	for _, dep := range c.Metadata.Dependencies {
+		if dep.Name != eSchema.SubChartName {
+			continue
+		}
+
+		if dep.Alias != "" {
+			eSchema.SubChartName = dep.Alias
+		}
+	}
+
+	return eSchema, nil
+}
+
+func GetEngineClusterDef(c *chart.Chart) (string, error) {
+	if c == nil {
+		return "", fmt.Errorf("failed to get the cluster definition of %s, chart is nil", c.Name())
+	}
+
+	clusterFilePath := filepath.Join(templatesDir, clusterFile)
+	for _, tpl := range c.Templates {
+		if tpl.Name != clusterFilePath {
+			continue
+		}
+
+		// get cluster definition from cluster.yaml
+		pattern := "  clusterDefinitionRef: "
+		str := string(tpl.Data)
+		start := strings.Index(str, pattern)
+		if start != -1 {
+			end := strings.IndexAny(str[start+len(pattern):], " \n")
+			if end != -1 {
+				return strings.TrimSpace(str[start+len(pattern) : start+len(pattern)+end]), nil
+			}
+		}
+	}
+	return "", fmt.Errorf("failed to find the cluster definition of %s", c.Name())
+}
+
+// ValidateValues validates the given values against the schema.
+func ValidateValues(schema *EngineSchema, values map[string]interface{}) error {
+	if schema == nil {
+		return nil
+	}
+
+	validateFn := func(s *spec.Schema, values map[string]interface{}) error {
+		if s == nil {
+			return nil
+		}
+		v := validate.NewSchemaValidator(s, nil, "", strfmt.Default)
+		err := v.Validate(values).AsError()
+		if err != nil {
+			// the default error message is like "cpu in body should be a multiple of 0.5"
+			// the "in body" is not necessary, so we remove it
+			errMsg := strings.Replace(err.Error(), " in body", "", -1)
+			return errors.New(errMsg)
+		}
+		return nil
+	}
+
+	if err := validateFn(schema.Schema, values); err != nil {
+		return err
+	}
+	return validateFn(schema.SubSchema, values)
+}
+
+// getEngineChartName gets the chart name for the given cluster engine type.
+func getEngineChartName(e EngineType) string {
+	eStr := strings.ToLower(string(e))
+	switch e {
+	case MySQL:
+		return "apecloud-" + eStr + "-cluster"
+	default:
+		return eStr + "-cluster"
+	}
+}
+
+func loadHelmChart(fs debme.Debme, name string) (*chart.Chart, error) {
+	file, err := fs.Open(name + ".tgz")
+	if err != nil {
+		return nil, err
+	}
+	defer file.Close()
+
+	c, err := loader.LoadArchive(file)
+	if err != nil {
+		if err == gzip.ErrHeader {
+			return nil, fmt.Errorf("file '%s' does not appear to be a valid chart file (details: %s)", name, err)
+		}
+	}
+
+	if c == nil {
+		return nil, fmt.Errorf("failed to load engine helm chart %s", name)
+	}
+	return c, err
+}
+
+func SupportedEngines() []EngineType {
+	return []EngineType{MySQL}
```

Comment:
no need to hard coded, I would write an `func init()` for each `go:embed` go module to register its own cluster type name.

Reply:
Use a separate go file `mysql.go` and call `init` to register ClusterType.